lib/search_files.js

```diff
@@ -0,0 +1,252 @@
+    if (ret) {
+      return ret;
+    }
+  }
+  return null;
+}
+
+function eachUnNullElem(arr, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  var ret = [];
+  for (var i = 0, n = arr.length; i < n; i++) {
+    var v = fn.apply(arr, [arr[i]].concat(args));
+    if (v) {
+      if (isArray(v)) {
+        Array.prototype.push.apply(ret, v);
+      } else {
+        ret.push(v);
+      }
+    }
+  }
+  return ret;
+}
+
+
+function normalizeHomedir(pth, homedir) {
+  if (!pth) {
+    return pth;
+  }
+
+  if (pth[0] === '~') {
+    if (pth.length === 1) {
+      return homedir;
+    }
+
+    if (pth[1] === '/') {
+      return path.join(homedir, pth.slice(2));
+    }
+
+    if (isWindows && pth[1] === '\\') {
+      return path.join(homedir, pth.slice(2));
+    }
+  }
+
+  return pth;
+}
+
+function normalizePath(pth, ctx) {
+  if (isFunction(pth)) {
+    pth = pth();
+  }
+
+  pth = normalizeHomedir(pth, ctx.homedir);
+
+  if (!isAbsolutePath(pth)) {
+    return path.join(ctx.cwd, pth);
+  }
+
+  return path.normalize(pth);
+}
+
+
+var initCtx = {
+  homedir: oshomedir(),
+  cwd: process.cwd(),
+};
+
+function normalizeContext(ctx) {
+  ctx = mapValues(ctx || {}, doNothing);
+
+  ctx.cwd = !ctx.cwd ? initCtx.cwd :
+    normalizePath(ctx.cwd, initCtx);
+
+  ctx.homedir = !ctx.homedir ? initCtx.homedir :
+    normalizePath(ctx.homedir, initCtx);
+
+  return ctx;
+}
+
+function normalizeExtensions(exts) {
+  if (isFunction(exts)) {
+    exts = exts();
+  }
+
+  if (isString(exts)) {
+    return [exts];
+  }
+
+  if (isArray(exts)) {
+    return exts.filter(isString);
+  }
+
+  if (isPlainObject(exts)) {
+    return Object.keys(exts).sort();
+  }
+
+  return [''];
+}
+
+function findFile(pth, basedir, findUp, opeArr) {
+
+  var parsed = parsePath(pth);
+  if (parsed.isAbsolute) {
+    findUp = false;
+    basedir = parsed.root;
+    pth = pth.slice(parsed.root.length);
+  } else if (parsed.root) {  // pth has a drive letter on windows.
+    basedir = path.resolve(parsed.root);
+    pth = pth.slice(parsed.root.length);
+  }
+
+  if (!findUp) {
+    pth = path.join(basedir, pth);
+    if (fs.existsSync(pth)) {
+      return opeArr([pth], checkSymbolicLink);
+    }
+    return opeArr([]);
+  }
+
+  var lastDir;
+  do {
+    var fp = path.join(basedir, pth);
+    if (fs.existsSync(fp)) {
+      return opeArr([fp], checkSymbolicLink);
+    }
+
+    lastDir = basedir;
+    basedir = path.dirname(basedir);
+  } while (lastDir !== basedir);
+
+  return opeArr([]);
+}
+
+function checkSymbolicLink(pth) {
+  if (!fs.lstatSync(pth).isSymbolicLink()) {
+    return pth;
+  }
+
+  try {
+    if (fs.existsSync(fs.realpathSync(pth))) {
+      return pth;
+    }
+  } catch (e) {}
+
+  return null;
+}
+
+
+function search(opts, ctx) {
+  ctx = normalizeContext(ctx);
+  return mapValues(opts || {}, searchWithOpt);
+
+  function searchWithOpt(opt, optname) {
+    if (!opt || isString(opt) || isArray(opt)) {
```

Comment:
(I'm sorry for such a long comment.)

Is the spec right as follows? 
(In such form, as you supposed above, it wouldn't be necessary to use `configFiles` option key as a part of filepath. In addition, I'd like to support user specified key which isn't used here but is inherited for being used after.)

```js
(setting)
configFiles = {
  rcFile: '.apprc',
  configs: ['~/config/.app', './.config/app.conf'],
  aaa: { path: 'path/to/config/file' },
  bbb: { path: ['path/to/file', '/root/path/to/file'], findup: true, exitEarly: true  },
  ccc: [
    { path: ['/root/path/to/file', 'path/to/file'], findup: true, exitEarly: true },
    { path: ['ddd/eee', 'fff/ggg'], exitEarly: true }
  ],
  hhh: { path: 'path/to/file', userSpecKey: 'iii' },
}
--
(expand, and process all in same way)
configFiles = {
  rcFile: [
    { path: '.apprc', findup: false, exitEarly: false },
  ],
  configs: [
    { path: '~/config/.app', findup: false, exitEarly: false },
    { path: './.config/app.conf', findup: false, exitEarly: false },
  ],
  aaa: [
    { path: 'path/to/config/file', findup: false, exitEarly: false },
  ],
  bbb: [
    { path: 'path/to/file', findup: true, exitEarly: true },
    { path: '/root/path/to/file', findup: true, exitEarly: true },
  ],
  ccc: [[
    { path: '/root/path/to/file', findup: true, exitEarly: true },
    { path: 'path/to/file', findup: true, exitEarly: true },
   ],[
    { path: 'ddd/eee', findup: false, exitEarly: true },
    { path: 'fff/ggg', findup: false, exitEarly: true },
  ]],
  hhh: [
    { path: 'path/to/file', findup: false, exitEarly: false, userSpecKey: 'iii' }
  ]
}
--
(result)
configFiles = {
  rcFile: [
    { path: '/root/path/to/cwd/.apprc' },
  ],
  configs: [
    { path: '/root/path/to/homedir/config/.app' },
    { path: '/root/path/to/cwd/.config/app.conf' },
  ],
  aaa: [
    { path: '/root/path/to/config/file' },
  ],
  bbb: [
    { path: '/find/up/from/cwd/path/to/file' },
  ],
  ccc: [
    { path: '/root/path/to/file' },
    { path: '/root/path/to/fff/ggg' }
  ],
  hhh: [
    { path: '/root/path/to/file', userSpecKey: 'iii' },
  ]
}
```